File: src/algorithm/random.sudoku.js
// // random row and col
var matRandom;

function isSave(i,col){
    // box
    let colBox = col - col%3
    for(let c = colBox; c < colBox+3; c++){
        if(matRandom[0][c] === i )return false
    }
    return true
}

function creatematRandomrix(){
    return Array(9).fill(0).map(() => Array(9).fill(0));
}

function swapLastCol(num,i,times){
    let j = 0
    while (true){
        if (isSave(num[i],j) && isSave(matRandom[times][j],8)){
            num[i] = matRandom[times][j]
            matRandom[times][j] = num[i]
            return 
        }
        j += 1
    }
}

function randomNumber(){
    let times = 0
    let n = 1
    while (times <= n){
        let queue = [];
        let num = [];
        // add 1-9 to queqe
        for(let i = 1 ; i < 10; i++) queue.push(i);
        // random number from queue 0 - 8 at first
        while (queue.length){
            let pos = Math.floor(Math.random()*queue.length)
            num.push(queue[pos])
            queue.splice(pos,1)
        }
        // insert num to first row of matRandomrix
        let i = 0
        for(let col = 0 ; col < 9; col++){   
            if(times > 0){
                while (!isSave(num[i],col)){
                    if (col === 6 || col === 7 || col === 8) {
                        swapLastCol(num,i,times);
                    }
                    else { // push and goto next i
                        num.push(num[i])
                        i += 1 
                    }
                }
            }
            matRandom[times][col] = num[i];
            i += 1  
        }
        times += 1
    }
}

export function randomSudoku(){
    matRandom = creatematRandomrix();
    randomNumber();
    return matRandom;
}

File: src/algorithm/random.sudoku.py
import random
# random row and col
mat = []
zeroList = []
maxZero = 0
maxMat = []

def isSave(i,col):
    # box
    colBox = col - col%3
    for c in range(colBox,colBox+3):
        if mat[0][c] == i:
            return False
    return True


def createMatrix():
    for i in range(9):
        mat.append([0 for x in range(9)])
    # print(mat)

def swapLastCol(num,i,times):
    j = 0
    while True:
        if isSave(num[i],j) and isSave(mat[times][j],8):
            num[i] = mat[times][j]
            mat[times][j] = num[i]
            return 
        j += 1

def randomNumber(level): # level [1,2,3,4,5]
    times = 0
    n = 1
    while times <= n:
        queue = [] 
        num =[]
        # add 1-9 to queqe
        for i in range(1,10):
            queue.append(i)
        # random number from queue 0 - 8 at first
        while len(queue):
            pos = random.randint(0,len(queue)-1)
            num.append(queue[pos])
            del queue[pos]
        # insert num to first row of matrix
        i = 0
        for col in range(9):
            if(times > 0):
                while not isSave(num[i],col):
                    if col in [6,7,8]:
                        swapLastCol(num,i,times)
                    else: 
                        num.append(num[i])
                        i += 1
            mat[times][col] = num[i]
            # print(num,num[i])
            i += 1  
        del num
        # print(mat[0])
        times += 1

def findZero():
    global maxZero
    global maxMat
    for row in range(9):
        colList = []
        for col in range(9):
            if mat[row][col] == 0:
                colList.append(col)
        zeroList.append(colList)
    a = 0
    for i in zeroList:
        a += len(i)
    if a > maxZero:
        maxZero = a
        maxMat = mat

def printArray():
    for i in mat:
        for j in i:
            print(j,end=' ')
        print()
t = 100
print(t)
maxZero = 0
while(t):
    mat = []
    zeroList = []
    createMatrix()
    randomNumber(5)
    # findZero()
    printArray()
    # print(*mat,sep='\n')
    t -= 1


# print(maxZero)
# print(*mat,sep = '\n')
